### PR TITLE
Fix with items parsing for yield, starred, named expr

### PR DESCRIPTION
**Pull request by dhruvmanila**
## Summary

This PR fixes with item parsing considering the yield, starred and named expression.

### But, what's the bug?

The yield and starred expressions weren't being considered when determining whether the with items are parenthesized or if it's a parenthesized expression. Taking a simple example:

```py
with (item1, *item2): ...
```

This is parsed as two individual with items currently but this is actually a tuple expression with two elements. This is easier to understand with the grammar:

```ebnf
with_stmt:
    | 'with' '(' ','.with_item+ ','? ')' ':' block 
    | 'with' ','.with_item+ ':' [TYPE_COMMENT] block 

with_item:
    | expression 'as' star_target &(',' | ')' | ':') 
    | expression 
```
(Omitted the `async` variant for brevity.)

Here, the PEG parser will _try_ the first variant in `with_stmt` which is a parenthesized with items and if that fails then it'll try the second variant. The grammar for a with item is `expression` which doesn't allow yield, starred, and named expression. This means that if they're present then the first variant would fail and it'll be the second variant which can possibly parse them.

### Solution

I've updated the logic so that it resembles what the PEG parser does.

Our algorithm starts in a similar manner by assuming that it's a parenthesized with items. This means that the `parse_with_item` should parse the context expression with a higher grammar rule between parenthesized with items and parenthesized expression. Here's an explanation mentioned in the code:

```rs
// If it's in an ambiguous state, the parenthesis (`(`) could be part of any
// of the following expression:
//
// Tuple expression          -  star_named_expression
// Generator expression      -  named_expression
// Parenthesized expression  -  (yield_expr | named_expression)
// Parenthesized with items  -  expression
//
// Here, the right side specifies the grammar for an element corresponding
// to the expression mentioned in the left side.
//
// So, the grammar used should be able to parse an element belonging to any
// of the above expression. At a later point, once the parser understands
// where the parenthesis belongs to, it'll validate and report errors for
// any invalid expression usage.
//
// Thus, we can conclude that the grammar used should be:
//      (yield_expr | star_named_expression)
```

Now, there are two important things to determine the with item kind:
1. If any of the with item has optional variables, it's a parenthesized with items
2. If any of the context expression is yield, starred or named, it's a parenthesized expression

Whatever the with item kind turns out to be, the parser needs to limit the expression because, as mentioned above, it parses using the highest common grammar.

This is basically what the change is.

## Test Plan

Added new valid and invalid test cases and verified the snapshots.

### CODE REVIEW

File: crates/ruff_python_parser/src/parser/statement.rs [hunk 1]
```diff
@@ -1443,40 +1432,51 @@ impl<'src> Parser<'src> {
...
+                    let err = match parsed_with_item.item.context_expr {
+                        Expr::Named(_) => ParseErrorType::UnparenthesizedNamedExpression,
+                        Expr::Starred(_) => ParseErrorType::StarredExpressionUsage,
+                        Expr::Yield(_) | Expr::YieldFrom(_) => {
+                            ParseErrorType::InvalidYieldExpressionUsage
+                        }
+                        _ => continue,
+                    };
+                    self.add_error(err, &parsed_with_item.item.context_expr);
+                }
```
Comment (dhruvmanila) [inline crates/ruff_python_parser/src/parser/statement.rs:1457]: One or more with item have an optional variable, so it's a parenthesized with items. Limit the parsed expression by reporting error for invalid expressions in this context.

File: crates/ruff_python_parser/src/parser/statement.rs [hunk 2]
```diff
@@ -1564,94 +1590,94 @@ impl<'src> Parser<'src> {
...
-
-                    // For example: `with (item := 10 as foo): ...`
-                    if self.at(TokenKind::As) {
-                        self.add_error(
-                            ParseErrorType::OtherError(
-                                "unparenthesized named expression cannot be used here".to_string(),
-                            ),
-                            named_expr.range(),
-                        );
-                    }
```
Comment (dhruvmanila) [inline crates/ruff_python_parser/src/parser/statement.rs:1598]: This check is now automatically done when limiting the expression in `parse_parenthesized_with_items`

File: crates/ruff_python_parser/src/parser/tests/snapshots/ruff_python_parser__parser__tests__parser__tests__parse_with_stmt.snap [hunk 3]
```diff
@@ -473,7 +473,7 @@ Program {
                         is_async: false,
                         items: [
                             WithItem {
-                                range: 238..245,
+                                range: 237..246,
                                 context_expr: Yield(
```
Comment (dhruvmanila) [inline crates/ruff_python_parser/src/parser/tests/snapshots/ruff_python_parser__parser__tests__parser__tests__parse_with_stmt.snap:477]: This range update is similar to the one for named expression:

```py
with (x := 1): ...
#    ^^^^^^^^ with item range
#     ^^^^^^ expression range

with (yield x): ...
#    ^^^^^^^^^ with item range
#     ^^^^^^^ expression range
```

File: crates/ruff_python_parser/tests/snapshots/valid_syntax@statement__ambiguous_lpar_with_items.py.snap
Comment (dhruvmanila) [inline crates/ruff_python_parser/tests/snapshots/valid_syntax@statement__ambiguous_lpar_with_items.py.snap:1]: The new test cases are added in their respective sections in the test file. But, I verified that the range updates are correct by temporarily moving all of the new test cases at the bottom of the file.

**Review by MichaReiser — APPROVED** [review]

**Pull request merged**